### DISCUSSION TITLE
Fallthrough sections

### DISCUSSION THREAD

**Discussion by jgonggrijp** [mustache/spec#178] [Ideas]
@determin1st has been championing a variant of this idea in several other discussions. For reference, I will link two posts where he described it: https://github.com/mustache/spec/discussions/157#discussioncomment-7598479 and before that, https://github.com/mustache/spec/issues/135#issuecomment-1776111072. The present discussion is dedicated to any feature that would give Mustache a concise notation for if/elif/elif.../else like fallthrough cases.

The long time status quo in Mustache is that you need to repeat a condition in order to do an if/else:

```
{{#likes_cats}}
Hey, I hear you like cats.
{{/likes_cats}}
{{^likes_cats}}
Hey, why don't you like cats?
{{/likes_cats}}
```

If you lean very much to the pragmatist side (#163), situations like the above make it tempting to introduce an `else` keyword into the language:

```
{{#likes_cats}}
Hey, I hear you like cats.
{{else}}
Hey, why don't you like cats?
{{/likes_cats}}
```

I feel this would be very un-Mustachy. `{{else}}` should just be an interpolations tag that looks for an `else` key in the context.

It gets a lot better already if you use a `|` sigil instead of an `else` keyword:

```
{{#likes_cats}}
Hey, I hear you like cats.
{{|}}
Hey, why don't you like cats?
{{/likes_cats}}
```

In combination with an inverted section, this notation also lets you do something that currently cannot be done without data preprocessing (see option 9 in #157):

```
{{^list}}
Rendered if list is empty (could be omitted).
{{|}}
Rendered *once* if list is nonempty.
{{/list}}
```

It's just that the use case is a bit narrow. However, if `|` is a sigil, that means there could potentially be some kind of payload after it. That would make it possible to extend this syntax to a Mustache equivalent of a switch statement with multiple fallthrough cases (note that this relies on #175 to avoid a confusing closing tag):

```
{{#likes_cats}}
Hey, I hear you like cats.
{{|likes_dogs}}
Hey, I hear you like dogs.
{{|}}
Hey, why don't you like pets?
{{/}}
```

We save ourselves three tags per case in this way. If we compare this to status quo Mustache, the saved effort is quite substantial even for just three cases:

```
{{#likes_cats}}
Hey, I hear you like cats.
{{/likes_cats}}
{{^likes_cats}}
    {{#likes_dogs}}
Hey, I hear you like dogs.
    {{/likes_dogs}}
    {{^likes_dogs}}
Hey, why don't you like pets?
    {{/likes_dogs}}
{{/likes_cats}}
```

I suspect this is the idea that @determin1st originally came up with and that eventually evolved into the feature he implemented in sm-mustache (see references above). The idea that he actually implemented is more intricate than what I illustrated here; it includes equality comparisons to literal values and multiple fallthrough cases within a single `|` tag. I personally think those intricacies deviate too much from Mustache's core principles, so I defer to @determin1st if he feels like discussing them.

Continuing from the above example, I see two problems. Firstly, it makes sense that `{{|` cases can be chained both after `{{#` and after `{{^`. It also makes sense that conditions in a `{{|` can be written in the same way as in a `{{#` or a `{{^`, so with dotted names that resolve to the context. However, this means that it is ambiguous whether the `{{|` itself is a positive (`#`) or a negative (`^`) condition. It is easy to imagine situations where one might want to chain a mixture of positive and negative cases, so it would be too restrictive to always use the sign of the `{{#` or `{{^` that opened the chain.

Secondly, the `|` could be confused for the `|` operator in filter syntax (#153), an idea I'm more fond of than fallthrough sections.

The first solution that I thought of, but which does not work, is to simply start a new `{{#` or `{{^` without closing the previous one. This does not work, because it is impossible for a parser to distinguish from nested sections with omitted closing tags. It also means that the `|` sigil is still required, just to mark the default case:

```
{{#likes_cats}}
Hey, I hear you like cats.
{{#likes_dogs}}
Hey, I hear you like dogs.
{{|}}
Hey, why don't you like pets?
{{/}}
```

Another option, which would work but which I dislike, would be to use double sigils:

```
{{#likes_cats}}
Hey, I hear you like cats.
{{|#likes_dogs}}
Hey, I hear you like dogs.
{{|}}
Hey, why don't you like pets?
{{/}}
```

Or yet another option that I would dislike: add a second sigil specifically for negative fallthrough cases, for example `{{\`.

I don't see an acceptable way yet in which fallthrough cases could be implemented. Despite that, the idea has grown on me, so I hope for another round of inspiration!

**Comment by determin1st** [comment]
> The idea that he actually implemented is more intricate than what I illustrated here; it includes equality comparisons to literal values and multiple fallthrough cases within a single `|` tag. I personally think those intricacies deviate too much from Mustache's core principles,

yes, `OR` sections (of the block, or subsections of the section) may contain __only pure values__ (stringifyed), not like standard `#` or `^` which are value fetchers by nature. it doesnt deviate much from `SWITCH` construct, common in programming languages, where the fallthrough is:

```js
switch (value) {
case 1:
  // ...
  // fallthrough to 2
case 2:
  // ...
  break;// no fallthrough
default:
  // ...
  break;
}
```

an absense of `break` which allows to partially execute code in both sections. im not sure how it can be utilized.
the `case` clause, is completely opposite - very useful and represents a [logical `OR` disjunction ](https://en.wikipedia.org/wiki/Logical_disjunction)
the only sigil that matches generic language OR operation is `|`.

the next Q to the public and implementers is whether stack pushing needed:
```
  {{#animal}}
    You favor "{{.}}" which Im unable to categorize.
    Do you mean {{.animal}}?
    What about {{animal}}?

  {{|cat}}
    You favor domesticated leisure species

  {{|dog}}
    You favor domesticated utility species

  {{|crocodile}}
    Are you really in favor of aquatic reptiles?

  {{/}}
```
with data
```php
[
  "animal" => "mouse"
]
```

in my implementation currently becomes:
```
    You favor "" which Im unable to categorize.
    Do you mean mouse?
    What about mouse?
```
do you agree with that behavior/logic?

EDIT: oke, i forgot that fallthrough is actually utilized by using this extended syntax:
```
{{#number}}
  non zero
{{|1|2|3|4.5}}
  number is either 1 or 2 or 3 or 4.5
{{|1}}
  number is 1 (never appears)
{{|}}
  zero
{{/}}
```

+EDIT
thus, the above mustachy `SWITCH` construct can be mapped to programming language as
```js
switch (number) {
case 1:
case 2:
case 3:
case 4.5:
  // ...
  break;
case 1:
  // ...
  break;
default:
  // ... non zero
  break;
inverted_default:
  // ... zero
  break;
}
```

**Reply by jgonggrijp** [reply]
Thanks for clarifying. 👍

I'm not fond of having a literal after the `|`, but I think I do understand why you chose to implement it this way. In the two-clause scenario, where we abbreviate `{{#x}}{{/x}}{{^x}}{{/x}}` to `{{#x}}{{|}}{{/x}}` or the same but with `#` and `^` reversed, both clauses concern the same variable. You generalized "all clauses must concern the same variable" to more than two clauses, which I find a reasonable rule. Since status quo Mustache only distinguishes between zero and non-zero, you had to come up with new syntax in order to enable more than two cases, and you chose equality comparisons to literal values. Does that correctly summarize your line of thought?

I now realize that in the opening post, I let go of the restriction that all clauses should be about the same variable. With hindsight, that is perhaps a bit too liberal. On the other hand, not having that restriction would play nice with the pattern I've had in mind for the "trailing comma + 1" problem. That is, if I have a list, say `['variables', 'sections', 'partials', 'lambdas']`, I want to concatenate it like `variables, sections, partials and lambdas`. Assuming I somehow already decorated the list elements so that each has boolean `first` and `last` properties, I need to write something like the following with status quo Mustache:

```
{{#list}}{{#last}} and {{/last}}{{^last}}{{^first}}, {{/first}}{{/last}}{{.}}{{/list}}
```

Fallthrough sections without a "same variable" rule would allow me to write something like this instead (though I still don't like the double sigil notation):

```
{{#list}}{{#last}} and {{|^first}}, {{/}}{{.}}{{/list}}
```

Regarding switch cases without a `break`, they are commonly used in a similar way to your `|1|2|3|4.5`:

```js
switch(number) {
case 1:
case 2:
case 3:
case 4.5:
    write('number is either 1 or 2 or 3 or 4.5');
case 0:
    break;
default:
    write('not zero');
}
```

This also illustrates a counter-intuitive side effect of the way you implemented the feature. Since you cannot start more specific than zero or non-zero, non-inverted section switches start with the default case instead of ending with it.

I think that, given a choice to restrict all cases to the same variable, you did the best possible, but I don't like it for the above reasons. I would also prefer to reserve a notation like `1|2|3|4.5` for #153.

**Reply by determin1st** [reply]
> Fallthrough sections without a "same variable" rule would allow me to write something like this instead (though I still don't like the double sigil notation):
> 
> ```
> {{#list}}{{#last}} and {{|^first}}, {{/}}{{.}}{{/list}}
> ```

such a syntax will not work for a single value in the list plus it would bring a mess into codebase and reduce template readability, the correct one is:
```
{{@list}}
  {{^_last}}
    {{^_first}}
      ,
    {{/}}
  {{|}}
    {{^first}}
      and
    {{/}}
  {{/}}
  {{.}}
{{/}}
```
i intentionally put `@` (block type which means iteration) and `_` (~~variable~~ path modifier which means special/assisted value related to iteration) as it currently implemented in my mustache. this variant can be also be implemented with lambdas i mentioned earlier.

the purpose of a `SWITCH` is to enable explicit branching based on value itself in the template - there are no other means capable of doing such a thing.


> This also illustrates a counter-intuitive side effect of the way you implemented the feature. Since you cannot start more specific than zero or non-zero, non-inverted section switches start with the default case instead of ending with it.

well, it's not a variable, it is a value which might be the result of a lambda invokation. variable `{{the.path.to.value}}` doesnt include sections (or subsections). although primary section stands for the default in `SWITCH`, it is allowed to write (JS/PHP/others):
```js
switch (value) {
default:
  break;
case 1:
  break;
}
```
only the order of conditional sections is taken into account.

> I think that, given a choice to restrict all cases to the same variable, you did the best possible, but I don't like it for the above reasons. I would also prefer to reserve a notation like `1|2|3|4.5` for #153.

if you want to standardize **some** syntax you need more examples that showcase its benefits.

**Reply by jgonggrijp** [reply]
Alright. I showed one realistic example, i.e., the trailing comma +1 scenario. How about you show a realistic scenario where your `{{|literal}}` syntax shines?

I mean, I understand that it mimics `switch` in C-like languages, but why would I need that in a template?

**Comment by determin1st** [comment]
> Alright. I showed one realistic example, i.e., the trailing comma +1 scenario. How about you show a realistic scenario where your `{{|literal}}` syntax shines?
> 
> I mean, I understand that it mimics `switch` in C-like languages, but why would I need that in a template?

any example that has more that two render variants of the same entity.

for example a `msg` here: https://github.com/determin1st/sm-utils/blob/main/tests/error-log.json
is the data example.

it can be:
- empty/absent (0)
- oneliner (path)(1)
- multiline text block (2)
- path + multiline (3)

so the analyzer creates `msg-type` which is switch cased in the template

render example
![photo_2023-09-10_15-09-45](https://github.com/mustache/spec/assets/16524081/a6db4578-a2c2-4ff9-bbd6-4e14fed63217)


github discussion is putting quotation at the top, which is wrong, but i wont fix it

**Reply by determin1st** [reply]
also, you can see the switching on error `level`: 0=green 1=yellow 2=red

**Reply by jgonggrijp** [reply]
Honestly, that example feels rather artificial to me. The example data were clearly written to showcase the feature.

Do you have an example where motivation would go in the other direction? By that, I mean that the data and the template needed to be written anyway, but your feature makes it much better.

**Reply by determin1st** [reply]
that's the real application of mustache - structured log renderer (btw i also didnt see any tree-like log formatter examples before). it will render both ANSI and plaintext variants. for ANSI, there will be colors, for plaintext, there will be tags. that's why `level` is a number, its better to store it like that.

again, the purpose of the `SWITCH` is branching, which is unattainable with current mustache spec or any other implementation.
whether im wrong, i suggest to repeat this template:
```
*
{{^level}}
  INFO
{{|1}}
  WARNING
{{|2}}
  ERROR
{{/}}
: {{message}}
```
with data:
```js
{
  level: 2,
  message: "something"
}
```
output
```
* ERROR: something
```

try to replace the template with current/better mustache, feeding the same data and getting the same output.
imo, this construct is obligatory in practical templating.

**Reply by jgonggrijp** [reply]
Thanks. By reducing the use case to its essence, you made it more convincing for me.

Current Mustache can do value-based branching using dynamic partials. I will readily admit that the following solution is not ideal, but it works:

```
{{! 0.mustache }}
INFO

{{! 1.mustache}}
WARNING

{{! 2.mustache }}
ERROR

{{! main.mustache }}
* {{>*level}}: {{message}}
```

[Playground](https://jgonggrijp.gitlab.io/wontache/playground.html) savestate (render the `main` template):

```
{"data":{"text":"{\n  level: 2,\n  message: \"something\"\n}"},"templates":[{"name":"0","text":"INFO"},{"name":"1","text":"WARNING"},{"text":"ERROR","name":"2"},{"name":"main","text":"* {{>*level}}: {{message}}"}]}
```

As I already acknowledged, this solution is not ideal. We need to have external templates with tiny integer names. The external templates themselves are also tiny. Despite that, the above solution also has two advantages over your syntax: it does not require writing literal values inside the template (see footnote), and it is easier to add more log levels later, or to change the labels of some of the log levels, because there is a single source of truth about the possible log levels.

The tiny external templates could be replaced by tiny *internal* templates using #162:

```
{{! main.mustache }}
{{:0}}INFO{{/0}}
{{:1}}WARNING{{/1}}
{{:2}}ERROR{{/2}}
* {{>*level}}: {{message}}
```

Catering to situations where the subtemplates are more than just a single word, we could alternatively address the tiny integer names using #153, so we could externalize the subtemplates with sensible names:


```js
{
     LEVEL_MAP: ['log_info', 'log_warn', 'log_error'],
     loglevel: function(level) {
         return LEVEL_MAP[level];
     }
}
```

```
{{! log_info.mustache }}
INFO

{{! log_warn.mustache }}
WARNING

{{! log_error.mustache }}
ERROR

{{! main.mustache }}
* {{>* level | loglevel }}: {{message}}
```

Both approaches could also be combined:

```
{{! main.mustache }}
{{:log_info}}INFO{{/log_info}}
{{:log_warn}}WARNING{{/log_warn}}
{{:log_error}}ERROR{{/log_error}}
* {{>* level | loglevel }}: {{message}}
```

In the general case, where the templates for the individual log levels might be complex, I think that a solution that combines dynamic partials with nested templates and/or filters is superior to what you are proposing. In this specific example, they are just single words, so I would probably return that word directly from a filter and skip the partial altogether:

```js
{
     LEVEL_MAP: ['INFO', 'WARNING, 'ERROR'],
     loglevel: function(level) {
         return LEVEL_MAP[level];
     }
}
```

```
{{! main.mustache }}
* {{ level | loglevel }}: {{message}}
```

All this being said, you are right that at least one new feature is needed. I would favor both #162 and #153 over your `{{|literal}}` syntax, because they also address other scenarios.

---

Footnote: literal values inside templates may seem innocent if they are just numbers or strings. However, literal values aren't always numbers or strings; they could also be entire datastructures. Having syntax for that is very far beyond the scope of Mustache, as far as I'm concerned.

**Reply by determin1st** [reply]
> Despite that, the above solution also has two advantages over your syntax: it does not require writing literal values inside the template (see footnote), and it is easier to add more log levels later, or to change the labels of some of the log levels, because there is a single source of truth about the possible log levels.

external templates? you mean that it is required to type a number as a filename? that would be 3 sources of truth (3 files) instead of one template. beside that, what if you need more branching on numeric values? like `msg-type` i mentioned before? wouldnt that be a name collision?

> The tiny external templates could be replaced by tiny _internal_ templates using #162:
> 
> ```
> {{! main.mustache }}
> {{:0}}INFO{{/0}}
> {{:1}}WARNING{{/1}}
> {{:2}}ERROR{{/2}}
> * {{>*level}}: {{message}}
> ```

yes, that is better, but still prone to name collision.
you also type number values into template. i dont see much benefit of `:1` in place of `|1`

> Catering to situations where the subtemplates are more than just a single word, we could alternatively address the tiny integer names using #153, so we could externalize the subtemplates with sensible names: ...
> 

here you realize the name collision *problem* and successfully workaround it.
the drawback is obscurity and cryptic (to me) syntax of `>* level | loglevel`

> In the general case, where the templates for the individual log levels might be complex, I think that a solution that combines dynamic partials with nested templates and/or filters is superior to what you are proposing. In this specific example, they are just single words, so I would probably return that word directly from a filter and skip the partial altogether:
> 
> ```js
> {
>      LEVEL_MAP: ['INFO', 'WARNING, 'ERROR'],
>      loglevel: function(level) {
>          return LEVEL_MAP[level];
>      }
> }
> ```
> 
> ```
> {{! main.mustache }}
> * {{ level | loglevel }}: {{message}}
> ```

one more step and mustache engine will be out of service :]
developer will write renderer functions in his language of choice.

to re-establish mustache authority, my implementation of lambdas comes into help:
```
{{#assign level-tag}}
  {{^level}}INFO
  {{|1    }}WARNING
  {{|2    }}ERROR
  {{/}}
{{/}}

* {{level-tag}}: {{message}}
```

rendering will depend on `assign` code (2 lines) but representational data will not scatter around like in your variant. this approach conforms the principle of separation of concerns.

> All this being said, you are right that at least one new feature is needed. I would favor both #162 and #153 over your `{{|literal}}` syntax, because they also address other scenarios.

internal templates (#162) doesnt need another invention, they are naturally appear inside sections (what i call blocks):
```
base template
{{#some}}
  internal template
  {{^another}}
    very internal template
  {{/}}
{{/}}
```
that's the case where less is more. less sigils, less logic, less mental load - more templating language

**Reply by jgonggrijp** [reply]
> external templates? you mean that it is required to type a number as a filename? that would be 3 sources of truth (3 files) instead of one template. beside that, what if you need more branching on numeric values? like `msg-type` i mentioned before? wouldnt that be a name collision?

Right, that is why I acknowledged that the current standard solution is not ideal. I think you are right that at least *some* new feature is needed in the Mustache language, in order to handle your use case in a satisfying way.

> 
> > The tiny external templates could be replaced by tiny _internal_ templates using #162:
> > (...)
> 
> yes, that is better, but still prone to name collision. you also type number values into template. i dont see much benefit of `:1` in place of `|1`

The difference is that `:1` names a template, not a literal value. The name just happens to consist of only a single digit. Yes, that might lead to naming collisions. I already acknowledged that this is not ideal.

> 
> > Catering to situations where the subtemplates are more than just a single word, we could alternatively address the tiny integer names using #153, so we could externalize the subtemplates with sensible names: ...
> 
> here you realize the name collision _problem_ and successfully workaround it. the drawback is obscurity and cryptic (to me) syntax of `>* level | loglevel`

Just before that, I showed you `{{>*level}}`. You thought it was fine, except for the potential name collisions due to the tiny integer template names (which, again, I already agreed with beforehand).

Now I'm adding `| loglevel`, and you think the syntax is cryptic. But the very syntax you are proposing to add looks like that, too!

> 
> > (...) In this specific example, [the templates for the individual log levels] are just single words, so I would probably return that word directly from a filter and skip the partial altogether:
> > (...)
> 
> one more step and mustache engine will be out of service :] developer will write renderer functions in his language of choice.

That example was not a "renderer function", but an extremely simple function that maps a number to a word. Contrary to the solution you showed before, the template author would not need to repeat herself if she needed to make the same mapping in more than one place. Which brings me to your new example, where you introduce lambda arguments and in-template context modification in order to solve that:

> 
> to re-establish mustache authority, my implementation of lambdas comes into help:
> 
> ```
> {{#assign level-tag}}
>   {{^level}}INFO
>   {{|1    }}WARNING
>   {{|2    }}ERROR
>   {{/}}
> {{/}}
> 
> * {{level-tag}}: {{message}}
> ```
> 
> rendering will depend on `assign` code (2 lines) but representational data will not scatter around like in your variant. this approach conforms the principle of separation of concerns.

Here I really disagree with you. So far, Mustache has made a conscious choice to leave some of the functionality in the host programming language, so that the template language can stay lean and mean. That, by itself, is a separation of concerns: let the host language do data processing and let the template language do the presentation.

You are basically proposing to turn Mustache itself into a programming language, which amounts to *removing* that separation of concerns.

> 
> > All this being said, you are right that at least one new feature is needed. I would favor both #162 and #153 over your `{{|literal}}` syntax, because they also address other scenarios.
> 
> internal templates (#162) doesnt need another invention, they are naturally appear inside sections (what i call blocks):
> 
> ```
> base template
> {{#some}}
>   internal template
>   {{^another}}
>     very internal template
>   {{/}}
> {{/}}
> ```

Sections do not replace internal templates, in the same way that loops do not replace functions.

**Reply by determin1st** [reply]
> The difference is that `:1` names a template, not a literal value. The name just happens to consist of only a single digit. Yes, that might lead to naming collisions. I already acknowledged that this is not ideal.

"If it walks like a duck and it quacks like a duck, then it must be a duck".
another difference is that mine has a strong cohesion (numbers map to level) and yours has a weak cohesion (numbers **may** map to level).
 
> Just before that, I showed you `{{>*level}}`. You thought it was fine, except for the potential name collisions due to the tiny integer template names (which, again, I already agreed with beforehand).

i ignored the syntax of a sloppy (error prone) variant because the correctness is what comes before clarity.
 
> Now I'm adding `| loglevel`, and you think the syntax is cryptic. But the very syntax you are proposing to add looks like that, too!

lets compare the clarity, your example adds:
- lambda/function which does mapping
- sigils `>`, `*` and `:`
- new names `log_info`,`log_warn`,`log_error` and `logLevel`

mine (the `SWITCH` block) is based on quite common `^` and `|` sigils. no lambdas and no tricky name mapping.

> That example was not a "renderer function", but an extremely simple function that maps a number to a word. Contrary to the solution you showed before, the template author would not need to repeat herself if she needed to make the same mapping in more than one place. 

i didnt say it was "renderer", i said developer will move to writing renderer functions where your `logLevel` will be a helper. as you summoned your lambdas to support your syntax variant, i summoned mine, it's fair comparison of `{{>*level | logLevel}}` vs `{{level-tag}}`.

> Here I really disagree with you. So far, Mustache has made a conscious choice to leave some of the functionality in the host programming language, so that the template language can stay lean and mean. That, by itself, is a separation of concerns: let the host language do data processing and let the template language do the presentation.

the concern is `INFO/WARNING/ERROR` which belongs to the view (the template).
if you put it into the code or the data, you'll break the principle.

if i wanted to break it, i wouldnt use `{{>*level | logLevel}}`, i would better mutated the data (or even better - created a helper), added `level-tag` key and used `{{level-tag}}`.

> You are basically proposing to turn Mustache itself into a programming language, which amounts to _removing_ that separation of concerns.

nah, my version effectively keeps `INFO/WARNING/ERROR` tags in the template (yes, in the runtime, it puts it into helper, but its a legal trick)

> Sections do not replace internal templates, in the same way that loops do not replace functions.

sections do not, but section's **content** is an internal template by nature.
re-inventing the wheel under another name, doesn't add value:
```
{{:path}}
  I'm an internal template
{{/}}
{{#path}}
  I'm NOT an internal template
{{/}}
```
it adds a confusion.

**Reply by jgonggrijp** [reply]
> lets compare the clarity, your example adds:
> 
> * lambda/function which does mapping
> * sigils `>`, `*` and `:`
> * new names `log_info`,`log_warn`,`log_error` and `logLevel`
> 
> mine (the `SWITCH` block) is based on quite common `^` and `|` sigils. no lambdas and no tricky name mapping.

I will admit that your syntax is simpler. However, it only addresses this use case and there is a large overlap with dynamic partials, which are already standard. While my syntax is perhaps not as clean, it relies on features that have wider applications.

Regarding `{{:path}}` and `{{#path}}`, you are right that both are subtemplates, but the difference is that the latter names a value in the context while the former names something that can be inserted using `{{>path}}`. They are in different namespaces. `{{$path}}` is a subtemplate as well, but it lives in the third namespace.

Besides that, `{{#path}}` has conditional and iterative semantics that are not helpful when you are writing the subtemplate in order to render it somewhere else using `{{>path}}`. You simply want it to never render at the place of definition.
